### PR TITLE
Guarin lig 539 multi gpu examples

### PR DESCRIPTION
**Pull request by guarin**
Adds example multi-gpu implementations for all models.

Example for Barlow Twins:
[Barlow Twins — lightly 1.2.2 documentation.pdf](https://github.com/lightly-ai/lightly/files/7898186/Barlow.Twins.lightly.1.2.2.documentation.pdf)

### CODE REVIEW

File: examples/pytorch_lightning_distributed/nnclr.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+class NNCLR(pl.LightningModule):
+    def __init__(self):
+        super().__init__()
+        resnet = torchvision.models.resnet18()
+        self.backbone = nn.Sequential(*list(resnet.children())[:-1])
+        self.projection_head = NNCLRProjectionHead(512, 512, 128)
+        self.prediction_head = NNCLRPredictionHead(128, 512, 128)
+        self.memory_bank = NNMemoryBankModule(size=4096)
+
+        self.criterion = NTXentLoss()
```
Comment (philippmwirth) [inline examples/pytorch_lightning_distributed/nnclr.py:23]: can we not gather here as well?

File: docs/source/examples/barlowtwins.rst [hunk 2]
```diff
@@ -27,3 +27,23 @@ Reference:
...
+            python lightly/examples/pytorch_lightning_distributed/barlowtwins.py
+
+        The model differs in the following ways from the non-distributed
+        implementation:
+
+        - Distributed Data Parallel is enabled
+        - Synchronized Batch Norm is used in place of standard Batch Norm
+        - Features are gathered from all GPUs before the loss is calculated
+
+        Note that Synchronized Batch Norm and feature gathering are optional and
```
[the remark this answers is not in the document]
Comment (guarin) [reply docs/source/examples/barlowtwins.rst:46]: You can still perform distributed training without Sync Batch Norm and without gathering features.

File: docs/source/examples/moco.rst [hunk 3]
```diff
@@ -34,3 +34,22 @@ Tutorials:
...
+        training with Pytorch Lightning. At least one GPU must be available on 
+        the system. The example can be run from the command line with::
+
+            python lightly/examples/pytorch_lightning_distributed/moco.py
+
+        The model differs in the following ways from the non-distributed
+        implementation:
+
+        - Distributed Data Parallel is enabled
+        - Synchronized Batch Norm is used in place of standard Batch Norm
```
[the remark this answers is not in the document]
Comment (guarin) [reply docs/source/examples/moco.rst:49]: No, it uses negatives from the memory bank instead.

File: examples/pytorch_lightning_distributed/nnclr.py (hunk 1 above)
Comment (guarin) [reply examples/pytorch_lightning_distributed/nnclr.py:23]: As in moco we are using a memory bank so it doesn't make sense to gather here.

File: docs/source/examples/barlowtwins.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (philippmwirth) [reply docs/source/examples/barlowtwins.rst:46]: Yes I meant that you could add one extra sentence telling the reader what happens if they choose to not use it.

File: docs/source/examples/moco.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (philippmwirth) [reply docs/source/examples/moco.rst:49]: true

File: docs/source/examples/barlowtwins.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (guarin) [reply docs/source/examples/barlowtwins.rst:46]: Ah ok :) Added it.

**Review by philippmwirth — APPROVED** [review]
